### PR TITLE
[2.3.1]Fix/2025 make elasticsearch ingest usable (required)

### LINKED ISSUE #2001: Limit default handled file for elasticsearch
## Bug description & symptom

Currently, we try to upload any file to elasticsearch with ingest mecanism.
We should avoid uploading any file or at least make this configurable.

## Diagnostic

Case we do not want to index file:
- Not document file: we can use whitelist_mimetype parameter but we should rely know value to put in it by default which is not trivial
- Big file. as we send file as b64, see #1877, it cause memory trouble and take time for large file. We should probably have some thing of limit for this. Another reason to have strict limit here is that it's almost certain that elasticsearch process itself as by default some kind of similar limitation, we should do thing in a way it work similar way (same limitation for elasticsearch ingest and tracim indexing). We should also think about making this configurable is possible.

## Expected

Unclear

### PR DESCRIPTION
**Pull request by inkhey**
related to #2025 .
this is a fix for #2001 and #2016 

## Checkpoints

The goal of this section is to help code reviewer to do required checks on this pull request. Please don't edit this list.
If one or more of these checkpoints are out of scope, please write below why they are.

- [x] Code is clear enough
- [x] If there are FIXME in the code, then there are associated issues and issue is referenced in the FIXME
- [x] If there are TODO, NOTE or HACK in code, date and developer initials are present
- [x] Automated tests have been written (covering feature or non regression), or *at least* an issue has been created for test implementation

### CODE REVIEW

**Comment by PhilippeAccorsi** [comment]
When we merge to "maintenance/2.3" we need to add this fix in develop.

File: backend/development.ini.sample [hunk 1]
```diff
@@ -420,6 +422,15 @@ search.engine = simple
 # content for only file with those mimetype, if set to empty, allow any mimetype, default value
 # allow most text-based content type.
 ; search.elasticsearch.ingest.mimetype_whitelist = text/markdown,text/plain
+# if you set ingest mimetype blacklist to list of separated mimetype by "," , you will not index file
+# content with those mimetype (even in included in whitelist), if set to empty, do nothing, default
+# value to "" meaning no blacklist.
+; search.elasticsearch.ingest.mimetype_blacklist = text/markdown
```
Comment (lebouquetin) [inline backend/development.ini.sample]: you should add a pertinent example like .zip or .tar.gz mimetypes

File: backend/tracim_backend/command/search.py [hunk 2]
```diff
@@ -36,9 +36,11 @@ def take_app_action(self, parsed_args: argparse.Namespace, app_context: AppEnvir
         self.search_api.create_index()
         print("Index template was created")
         if parsed_args.index_all:
-            print("Indexing all content")
-            self.search_api.index_all_content()
-            print("All content where indexed")
+            nb_errors = self.search_api.index_all_content()
+            if nb_errors == 0:
+                print("All content where indexed")
```
Comment (lebouquetin) [inline backend/tracim_backend/command/search.py]: you should write the number of indexed contents.

File: backend/tracim_backend/lib/search/search.py [hunk 3]
```diff
@@ -357,20 +378,97 @@ def index_content(self, content: ContentInContext) -> None:
...
+                'Content file of content "{}" will be not indexed: ingest mode disabled'.format(
+                    content.content_id
+                ),
+            )
+            return False
+
+        if not content.content.depot_file or content.size is None:
+            logger.debug(
+                self,
+                'Content file of content "{}" will be not indexed:  not a correct file format'.format(
```
Comment (lebouquetin) [inline backend/tracim_backend/lib/search/search.py]: I suggest to replace `'Content file of content "{}" will be not indexed:  not a correct file format'` by `Binary content of content "{}" not indexed:  invalid file format'`

File: backend/tracim_backend/lib/search/search.py [hunk 4]
```diff
@@ -357,20 +378,97 @@ def index_content(self, content: ContentInContext) -> None:
...
+
+        # INFO - G.M - 2019-06-24 - check mimetype validity
+        if (
+            self._config.SEARCH__ELASTICSEARCH__INGEST__MIMETYPE_WHITELIST
+            and content.mimetype
+            not in self._config.SEARCH__ELASTICSEARCH__INGEST__MIMETYPE_WHITELIST
+        ):
+            logger.debug(
+                self,
+                'Content file of content "{}" will be not indexed: mimetype "{}" not allowed to be indexed (whitelist rule)'.format(
```
Comment (lebouquetin) [inline backend/tracim_backend/lib/search/search.py]: Replace `'Content file of content "{}" will be not indexed: mimetype "{}" not allowed to be indexed (whitelist rule)'`
by `'Skip binary indexation of content "{}": mimetype "{}" is not whitelisted'`

File: backend/tracim_backend/lib/search/search.py [hunk 5]
```diff
@@ -357,20 +378,97 @@ def index_content(self, content: ContentInContext) -> None:
...
+
+        # INFO - G.M - 2019-06-24 - check mimetype validity
+        if (
+            self._config.SEARCH__ELASTICSEARCH__INGEST__MIMETYPE_BLACKLIST
+            and content.mimetype in self._config.SEARCH__ELASTICSEARCH__INGEST__MIMETYPE_BLACKLIST
+        ):
+            logger.debug(
+                self,
+                'Content file of content "{}" will be not indexed: mimetype "{}" not allowed to be indexed (blacklist rule)'.format(
+                    content.content_id, content.mimetype
```
Comment (lebouquetin) [inline backend/tracim_backend/lib/search/search.py]: 'Skip binary indexation of content "{}": mimetype "{}" is blacklisted'

File: backend/tracim_backend/lib/search/search.py [hunk 6]
```diff
@@ -357,20 +378,97 @@ def index_content(self, content: ContentInContext) -> None:
...
+                ),
+            )
+            return False
+
+        if content.size == 0:
+            logger.debug(
+                self,
+                'Content file of content "{}" will be not indexed:  empty'.format(
```
Comment (lebouquetin) [inline backend/tracim_backend/lib/search/search.py]: => 'Skip binary indexation of content "{}": empty file'

File: backend/tracim_backend/lib/search/search.py [hunk 7]
```diff
@@ -357,20 +378,97 @@ def index_content(self, content: ContentInContext) -> None:
...
+                    content.content_id
+                ),
+            )
+            return False
+
+        # INFO - G.M - 2019-06-24 - check content size
+        if content.size > self._config.SEARCH__ELASTICSEARCH__INGEST__SIZE_LIMIT:
+            logger.debug(
+                self,
+                'Content file of content "{}" will be not indexed: size "{}" is  bigger than size limit ({})'.format(
```
Comment (lebouquetin) [inline backend/tracim_backend/lib/search/search.py]: 'Skip binary indexation of content "{}": binary is oversized ({} > {})'

File: backend/development.ini.sample (hunk 1 above)
Comment (inkhey) [reply backend/development.ini.sample]: done

File: backend/tracim_backend/command/search.py (hunk 2 above)
Comment (inkhey) [reply backend/tracim_backend/command/search.py]: done.

File: backend/tracim_backend/lib/search/search.py (hunk 3 above)
Comment (inkhey) [reply backend/tracim_backend/lib/search/search.py]: done.

File: backend/tracim_backend/lib/search/search.py (hunk 4 above)
Comment (inkhey) [reply backend/tracim_backend/lib/search/search.py]: done.

File: backend/tracim_backend/lib/search/search.py (hunk 6 above)
Comment (inkhey) [reply backend/tracim_backend/lib/search/search.py]: done.

File: backend/tracim_backend/lib/search/search.py (hunk 7 above)
Comment (inkhey) [reply backend/tracim_backend/lib/search/search.py]: done.

**Comment by PhilippeAccorsi** [comment]
I have tested populate with default parameter value on mirror of production and populate working good with file not bigger than 50Mo.

Populate take 5min.

**Comment by PhilippeAccorsi** [comment]
@inkhey can you add also parameter about timeout andingest size limit to documentation?

**Comment by inkhey** [comment]
> @inkhey can you add also parameter about timeout and ingest size limit to documentation?

@PhilippeAccorsi That's added to settings.md file, do you see anything else ?

**Comment by PhilippeAccorsi** [comment]
Sorry about ingest.
I not see parameter about timeout in setting.md

File: backend/development.ini.sample [hunk 8]
```diff
@@ -420,6 +422,15 @@ search.engine = simple
 # content for only file with those mimetype, if set to empty, allow any mimetype, default value
 # allow most text-based content type.
 ; search.elasticsearch.ingest.mimetype_whitelist = text/markdown,text/plain
+# if you set ingest mimetype blacklist to list of separated mimetype by "," , you will not index file
+# content with those mimetype (even in included in whitelist), if set to empty, do nothing, default
+# value to "" meaning no blacklist.
+; search.elasticsearch.ingest.mimetype_blacklist = text/markdown,application/gzip,application/zip
```
Comment (lebouquetin) [inline backend/development.ini.sample]: could you remove text/markdown from the example? It's not an expected use case to activate plain-text search and to exclude markdown documents.

File: backend/tracim_backend/command/search.py [hunk 9]
```diff
@@ -9,7 +9,35 @@
 from tracim_backend.models.context_models import ContentInContext
 
 
-class SearchIndexInitCommand(AppContextCommand):
+class IndexingCommand(AppContextCommand):
+    def _index_one_content(self, content_id: int) -> None:
```
Comment (lebouquetin) [inline backend/tracim_backend/command/search.py]: this command should be in the elasticsearch api.

**Pull request merged**